cleanlab/outlier.py

```diff
@@ -0,0 +1,534 @@
+             If `k` is not provided, then by default ``k = knn.n_neighbors`` or ``k = 10`` if ``knn is None``.
+             If an existing ``knn`` object is provided, you can still specify that outlier scores should use
+             a different value of `k` than originally used in the ``knn``,
+             as long as your specified value of `k` is smaller than the value originally used in ``knn``.
+       *  t : int, default=1
+             Optional hyperparameter only for advanced users.
+             Controls transformation of distances between examples into similarity scores that lie in [0,1].
+             The transformation applied to distances `x` is `exp(-x*t)`.
+             If you find your scores are all too close to 1, consider increasing `t`,
+             although the relative scores of examples will still have the same ranking across the dataset.
+
+     If `pred_probs` is passed in during ``fit()``, `params` could contain following keys:
+       *  confident_thresholds: np.ndarray, default = None
+             An array of shape ``(K, )`` where K is the number of classes.
+             Confident threshold for a class j is the expected (average) "self-confidence" for that class.
+             If you specify ``confident_thresholds`` here there is no need to call ``fit()`` before calling ``score()``.
+       *  adjust_pred_probs : bool, True
+             Account for class imbalance in the label-quality scoring by adjusting predicted probabilities
+             via subtraction of class confident thresholds and renormalization.
+             Set this to ``False`` if you prefer to skip accounting for class-imbalance. You do not have to pass in `labels` in that case.
+             See `Northcutt et al., 2021 <https://jair.org/index.php/jair/article/view/12125>`_.
+       *  method : {"entropy", "least_confidence"}, default="entropy"
+             OOD scoring method.
+
+             Letting ``P = pred_probs[i]`` denote the given predicted class-probabilities
+             for datapoint *i*, its score can either be:
+
+             - ``'entropy'``: ``- sum_{j} P[j] * log(P[j])``
+             - ``'least_confidence'``: ``1 - max(P)``
+
+    """
+
+    OUTLIER_PARAMS = {"k", "t", "knn"}
+    OOD_PARAMS = {"confident_thresholds", "adjust_pred_probs", "method"}
+    DEFAULT_PARAM_DICT: Dict[Union[str, int, None], Union[str, int, None, np.ndarray]] = {
+        "k": None,  # ood features param
+        "t": 1,  # ood features param
+        "knn": None,  # ood features param
+        "adjust_pred_probs": True,  # ood pred_probs param
+        "method": "entropy",  # ood pred_probs param
+        "confident_thresholds": None,  # ood pred_probs param
+    }
+
+    def __init__(self, params: dict = {}):
+        self._assert_valid_params(params, self.DEFAULT_PARAM_DICT)
+        self.params = self.DEFAULT_PARAM_DICT
+        if params is not None:
+            self.params = {**self.params, **params}
```

Comment:
Get rid of the `params is not None` check, since we expect empty dict `{}` as the default. We always want to execute the below update since it's fast and ensures we have safely copied the dict. 

We always want to be extremely cautious about assigning dicts to other variables without having created a deepcopy of them.